plugin/core/sessions.py

```diff
@@ -278,50 +279,189 @@ def clear_dotted_value(current: dict, dotted: str) -> None:
+        """
         pass
 
     @classmethod
-    def standard_configuration(cls) -> ClientConfig:
-        raise NotImplementedError()
-
-    @classmethod
-    def adjust_configuration(cls, configuration: ClientConfig) -> ClientConfig:
-        return configuration
+    def adjust_user_configuration(cls, configuration: ClientConfig) -> None:
+        pass
 
     @classmethod
     def can_start(cls, window: sublime.Window, initiating_view: sublime.View,
                   workspace_folders: List[WorkspaceFolder], configuration: ClientConfig) -> Optional[str]:
+        """
+        Determines ability to start. This is called after needs_update_or_installation and after install_or_update.
+        So you may assume that if you're managing your server binary, then it is already installed when this
+        classmethod is called.
+
+        :param      window:             The window
+        :param      initiating_view:    The initiating view
+        :param      workspace_folders:  The workspace folders
+        :param      configuration:      The configuration
+
+        :returns:   A string describing the reason why we should not start a language server session, or None if we
+                    should go ahead and start a session.
+        """
         return None
 
-    def on_initialized(self) -> None:
-        pass
+    def __init__(self, weaksession: 'weakref.ref[Session]') -> None:
+        """
+        Constructs a new instance.
 
-    def on_shutdown(self) -> None:
-        pass
+        :param      weaksession:  A weak reference to the Session. You can grab a strong reference through
+                                  self.weaksession(), but don't hold on to that reference.
+        """
+        self.weaksession = weaksession
+
+
+_plugins = {}  # type: Dict[str, Type[AbstractPlugin]]
+
+
+def register_plugin(plugin: Type[AbstractPlugin], update_global_configs: bool = True) -> None:
+    global _plugins
+    global client_configs
+    try:
+        config = ClientConfig(
+            name=plugin.name(),
+            binary_args=plugin.command(),
+            languages=plugin.languages(),
+            tcp_port=plugin.tcp_port(),
+            enabled=True,
+            init_options=plugin.initialization_options(),
+            settings=plugin.default_settings(),
+            env=plugin.env()
+        )
+        client_configs.add_external_config(config)
+        _plugins[config.name] = plugin
+        if update_global_configs:
+            client_configs.update_configs()
+    except Exception as ex:
+        exception_log("Failed to register plugin", ex)
+
+
+def unregister_plugin(plugin: Type[AbstractPlugin]) -> None:
+    global _plugins
+    global client_configs
+    try:
+        name = plugin.name()
+        client_configs.remove_external_config(name)
+        _plugins.pop(name, None)
+    except Exception as ex:
+        exception_log("Failed to unregister plugin", ex)
+    finally:
+        client_configs.update_configs()
```

Comment:
Is it easy to let plugin unregister itself now?
That would be needed as when plugin wants to update the server, sessions should be destroyed first so that file handles are freed (important on Windows).

Reply:
This is a good point. I will have to investigate this.

Reply:
This is https://github.com/sublimelsp/LSP/issues/899 and I'm not fixing it in this PR.

Reply:
OK, I'll see what I can do in this PR. It doesn't seem that much more work after all. And we might as well break all the things in one go.

Reply:
This turns out to be way harder than I thought. It's a gigantic mess of global state that isn't properly RAII-ed in the right places. Let's leave it for another PR.